model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/role/JpaRoleModelCriteriaBuilder.java

```diff
@@ -82,11 +86,29 @@ public JpaRoleModelCriteriaBuilder compare(SearchableModelField<? super RoleMode
                 }
             case IN:
                 if (modelField.equals(SearchableFields.ID)) {
-                    validateValue(value, modelField, op, String.class);
+                    if (value == null || value.length == 0) throw new CriterionNotSupportedException(modelField, op);
+
+                    final Collection collectionValues;
+                    if (value.length > 1) { // array
+                        collectionValues = Arrays.asList(value);
+                    } else if (value[0] instanceof Collection) {
+                        collectionValues = (Collection) value[0];
+                    } else if (value[0] instanceof Stream) {
+                        Stream<Object> str = ((Stream) value[0]);
+                        collectionValues = str.collect(Collectors.toCollection(LinkedList::new));
+                    } else {
+                        throw new CriterionNotSupportedException(modelField, op);
+                    }
 
                     return new JpaRoleModelCriteriaBuilder((cb, root) -> {
                         In<String> in = cb.in(root.get("id"));
-                        ((Stream<String>) value[0]).forEach(id -> in.value(id));
+                        for (Object id : collectionValues) {
+                            if (id instanceof String) {
+                                in.value((String) id);
+                            } else {
+                                throw new CriterionNotSupportedException(modelField, op);
+                            }
```

Comment:
Could you just use `Objects.toString(id, null)` rather than throwing an exception?

Reply:
I had to switch it back to throwing CriterionNotSupportedException as there is `UUID` type expected. Please let me know if it requires some update.